compiler/rustc_data_structures/src/stable_hasher.rs

```diff
@@ -161,6 +161,10 @@ impl Hasher for StableHasher {
         // 8 bytes. Since this prefix cannot occur when we hash a single byte, when we hash two
         // `isize`s that fit within a different amount of bytes, they should always produce a different
         // byte stream for the hasher.
+        //
+        // To ensure that this optimization hashes the exact same bytes on both little-endian and
+        // big-endian architectures, we compare the value with 0xFF before we convert the number
+        // into a unified representation (little-endian).
```

Comment:
The comment is correct but I think we could put more emphasis that the endianness conversion must be the last step because that creates platform-dependent *values* to get platform-independent *bytes*.

It would be clearer if `siphasher::write` were generic over `[u8; N]` instead of taking different primitives. Oh well.

Reply:
Well to be fair it contains an optimized implementation for these primitives, so it's probably worth it.
Should I add something like
```
First, we have to compare the value (which has to be done in a platform-dependent manner) and only then can we convert the number to the little-endian format (to ensure platform-independent bytes being hashed).
```
?

Reply:
> (which has to be done in a platform-dependent manner)

That's probably confusing. We're going from [platform-dependent byte-representation, platform-independent value] to [platform-independent byte-representation, platform-dependent value]. Which means all operations that depend on the value must happen before that and afterwards we could only do bit-twiddling operations.
It would be more obvious if we used `to_le_bytes`.

I don't mean to explain endianness, it's just about which things must be be done before and after the conversion. That's what I didn't consider during the review. 😓

Reply:
I find `.to_le()` and `.to_be()` to be really confusing and always use `to_le_bytes()` and `to_be_bytes()` instead, which makes it much less likely to get things accidentally wrong (by converting twice for example).

Now that we have const generics it would probably be easy to just change `SipHasher128::short_write()` to `SipHasher128::short_write<const LEN: usize>(&mut self, bytes: &[u8; LEN])`.

Reply:
That sounds like a good plan. Since this is a portability bug let's fix it first and then improve the design.